lib/vector-core/buffers/src/lib.rs

```diff
@@ -205,3 +217,45 @@ impl<T, S: Sink<T> + Unpin> Sink<T> for DropWhenFull<S> {
         self.project().inner.poll_close(cx)
     }
 }
+
+// Instrumenting events received by the memory buffer can be accomplished by
+// hooking into the lifecycle of writing events to the buffer, hence the
+// InstrumentMemoryBuffer wrapper. This is not necessary for disk buffers
+// which are instrumented for this at a lower level in their implementation.
+#[pin_project]
+pub struct InstrumentMemoryBuffer<S> {
```

Comment:
Have you checked if this wrapper impacts sink performance? This will be quite hot-path code, so we should be careful here.

Reply:
I haven't yet, thanks for the heads up. This'll be a good time to explore our benchmarking system. Will update here.

Reply:
From manually benchmarking against `master`, performance does appear somewhat impacted. I ran `time target/release/vector --config` [config](https://gist.github.com/001wwang/fdbff5d62e61fe94f6ba93a27b42c48b) (info log level enabled). Over 10 runs, this branch averaged `11.001 sec` in clock time and `master` averaged `10.634 sec`. Not the most rigorous test, but curious the threshold for a significant regression.

Waiting for CI benchmarks to re-run as I re-ran the test-suite to check a failure.

Reply:
Checked out `bench-transforms` and `bench` in the benchmark suite (as I believe buffer code performance changes should be easiest to determine there) and there are some regressions highlighted. However, they're on the nanosecond/microsecond time scale so not sure if that's expected given the added code. Let me know if there are other things I should look into. So far, based on the results, I think the instrumentation adds minimal overhead.

Reply:
Thanks for checking out the change in performance here! I also tried running the ad-hoc benchmark I usually run which did show some regression:

Before (2e1b3b609):

```
perf stat -r 10 bash -c 'yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector --config /tmp/tmp.toml 2> /dev/null'

 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector --config /tmp/tmp.toml 2> /dev/null' (10 runs):

      14850.405747      task-clock:u (msec)       #    1.703 CPUs utilized            ( +-  0.21% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
             10418      page-faults:u             #    0.702 K/sec                    ( +-  5.27% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

       8.721963605 seconds time elapsed                                          ( +-  0.25% )
```

After (401de5faf):

```
➜  vector git:(instrument-buffer-basics) ✗ perf stat -r 10 bash -c 'yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector --config /tmp/tmp.toml 2> /dev/null'

 Performance counter stats for 'bash -c yes "{\"appname\":\"su\",\"hostname\":\"initech.io\",\"message\":\"TPS report missing cover sheet\",\"msgid\":\"ID81\",\"procid\":4015,\"timestamp\":\"2020-12-19T21:48:09.004Z\",\"version\":3}" | head -n 1000000 | ./target/release/vector --config /tmp/tmp.toml 2> /dev/null' (10 runs):

      22122.551198      task-clock:u (msec)       #    1.734 CPUs utilized            ( +-  0.43% )
                 0      context-switches:u        #    0.000 K/sec
                 0      cpu-migrations:u          #    0.000 K/sec
              9459      page-faults:u             #    0.428 K/sec                    ( +-  2.09% )
   <not supported>      cycles:u
   <not supported>      instructions:u
   <not supported>      branches:u
   <not supported>      branch-misses:u

      12.755038089 seconds time elapsed                                          ( +-  0.45% )
```

The config was similar to yours:

```toml
[sources.gen]
type = "stdin"

[sinks.blackhole_testname]
# default, blocking memory buffer
type = "blackhole"
inputs = ["gen"]
```

I imagine the soak tests would show this too if there were a way to easily run a branch (cc/ @tobz for idea fodder).

Given this, it does seem like it'd be worth looking into profiling and seeing where the hot spots are. You could also try introducing batching to see the impact it has.

I think you could leverage the buffer-specific benchmarks: https://github.com/vectordotdev/vector/tree/master/lib/vector-core/buffers/benches to iterate on the performance. You'll need to wire up a metrics recorder though by calling https://github.com/vectordotdev/vector/blob/2e1b3b609af8519c36057afc90107f066e554bdc/lib/vector-core/src/metrics/mod.rs#L103-L110 or maybe just wiring up recorder in `buffers/benches` (to avoid the additional dependency) if that shows the same regression in performance.

Reply:
Awesome, thanks for the detail! Will go ahead and set this up.

Reply:
The roughly 50% increase in time taken shown by @jszwedko is concerning. I would expect that to have a significant impact on end-to-end performance. Would it make sense to punt on instrumenting the memory buffers and just merge the disk buffer bits, moving that work into a separate issue, in order to move forward in the issue at least partially?

Reply:
I'm not 100% sure if it is this wrapper or simply emitting metrics here in general that is causing the performance degradation. We could test it it with the wrapper but without emitting metrics to see what the overhead of the wrapper specifically is.

Reply:
Based on benchmarking with the `buffers/benches/in_memory` target (set up a `DebuggingRecorder` from `metrics-util` and tracing context layer to avoid shifting lots of dependencies with `VectorRecorder`), the `InstrumentMemoryBuffer` wrapper doesn't cause significant regressions. We see regressions when metrics are actually emitted. For example,
```
# wrapper, no instrumentation, baseline = master (fe92f720)
buffer-memory/write-then-read/4104                                                                          
                        time:   [3.5509 s 3.6023 s 3.6631 s]
                        thrpt:  [272.99 Kelem/s 277.60 Kelem/s 281.62 Kelem/s]
                 change:
                        time:   [-2.1557% -0.1921% +1.8202%] (p = 0.81 > 0.05)
                        thrpt:  [-1.7877% +0.1925% +2.2032%]

# wrapper, with instrumentation, baseline = master (fe92f720)
buffer-memory/write-then-read/4104                                                                          
                        time:   [4.3061 s 4.3877 s 4.5020 s]
                        thrpt:  [222.12 Kelem/s 227.91 Kelem/s 232.23 Kelem/s]
                 change:
                        time:   [+18.714% +21.568% +25.398%] (p = 0.00 < 0.05)
                        thrpt:  [-20.254% -17.742% -15.764%]
                        Performance has regressed.
```

Note, I locally changed the `max_events` in the benchmarks to `1_000_000` (and samples down to `10` for speed) as at `1_000` events, the results were quite noisy.

Comparing [master's flamegraph](https://drive.google.com/file/d/1_22q4Hrdv-Dv9e3Lgx6AZOPF-kOXTbdt/view?usp=sharing) and [this branch's flamegraph](https://drive.google.com/file/d/1OEbkxOMHodVDDC9Ou_Z0J4U7NUjQEDm-/view?usp=sharing) running the[ same config](https://gist.github.com/001wwang/fdbff5d62e61fe94f6ba93a27b42c48b), it's clear that buffer instrumentation does seem to takes up a solid chunk of the execution. I think it makes sense to go ahead and try the batching optimization. 

cc: @jszwedko, @bruceg

Reply:
Nice investigation @001wwang ! Your findings make sense to me.

Reply:
Update based on latest changes (4d9236f).

To improve memory buffer instrumentation performance, we update a set of atomics on read/write and emit metrics on an interval rather than emit metrics at each read/write. Looking at a `master (0a6a1865)` [flamegraph](https://drive.google.com/file/d/1IcDaPNE-zUCUSVsdSEdtaFmaffCDCmUL/view?usp=sharing) versus this branch's [flamegraph](https://drive.google.com/file/d/1GVgZAkm9aIc6-6t-lDWp_SS8qMfki2YB/view?usp=sharing) running the same config, there is a clear improvement over the original instrumentation method (no large chunk of execution spent emitting buffer metrics). 

Unfortunately, from running the `buffers/benches/in_memory` benchmarks (note: now requires introducing a tokio runtime when building buffers due to the new tokio task), it seems there are still significant regressions (will retry this). That said, from a simple end-to-end benchmark (run using [hyperfine](https://github.com/sharkdp/hyperfine) with default options) [results](https://gist.github.com/001wwang/550352730d40b90b0d283d6d8211820a) seem in line with master. 

For disk buffers, the original instrumentation strategy (emit on read/write) is, based on [similar simple end-to-end benchmarks](https://gist.github.com/001wwang/22bb8caee6540924a0fa647fe9103bcf), in line with master as well. I have yet to run the full benchmarks in `buffers/benches/on_disk`, but I think the disk buffer performance is limited enough by other factors that instrumentation will not cause significant regressions (a lean run with only `10` samples did not show any).

Moving forward, a couple questions:
1. Does this strategy for memory buffer instrumentation make sense? And, would we like to use the same model for disk buffers or is the direct emission simpler and easier to maintain (and disk buffers already benefit from their natural batching behavior)? I like this model as it works well with adding the `BufferCreated` event for initial data.

2. As Bruce mentioned earlier and now that we have more information, would we be interested in merging disk buffer instrumentation and splitting (or punting on) memory buffer instrumentation into its own PR? This PR is getting a bit long and messy, apologies.

cc: @jszwedko

Reply:
> Moving forward, a couple questions:
> 
>     Does this strategy for memory buffer instrumentation make sense? And, would we like to use the same model for disk buffers or is the direct emission simpler and easier to maintain (and disk buffers already benefit from their natural batching behavior)? I like this model as it works well with adding the BufferCreated event for initial data.

My preference would be to rely on a uniform implementation, one that collects on a timer like you've got in the disk buffer. I strongly suspect if you pushed this you'll see the memory benchmarks improve. Avoiding work per-event is usually a good move, and I think your results from the disk buffer bear that out. The calculus would change here if/when metrics evolves handles, save that we'll always pay the cost of logging overhead per-event.

>     As Bruce mentioned earlier and now that we have more information, would we be interested in merging disk buffer instrumentation and splitting (or punting on) memory buffer instrumentation into its own PR? 

I vote no on punting, unless we plan to punt on the whole thing but I vote no on that too. :) I am luke-warm on splitting the work across two PRs, unless we were confident we could get the next PR wrapped prior to the next release. From the point of view of a user it's a bit weird to have important telemetry for one kind of buffer but not another, is my thinking. 

If you wanted to slim this PR out you could migrate your benchmark improvements to another, smaller PR and land that first.  Possibly also the movement of tracing spans down into the buffer code could be a separate PR. 

> This PR is getting a bit long and messy, apologies.

I think we've learned a good deal about what we want to get out of this telemetry and how we make it feasible in the process, so it seems like we're moving in the right direction to me.